File: 0214-shortest-palindrome/0214-shortest-palindrome.java
class Solution {
    /**
     * Find the shortest palindrome by adding characters to the beginning of string s.
     * Uses rolling hash to find the longest palindrome prefix efficiently.
     * 
     * @param s the input string
     * @return the shortest palindrome formed by adding characters to the beginning
     */
    public String shortestPalindrome(String s) {
        // Rolling hash parameters
        final int BASE = 131;           // Base for polynomial rolling hash
        final int MOD = 1000000007;     // Large prime modulus to avoid overflow
      
        // Hash values for prefix and suffix
        int prefixHash = 0;              // Hash of prefix (forward direction)
        int suffixHash = 0;              // Hash of suffix (reverse direction)
      
        // Multiplier for suffix hash calculation
        int multiplier = 1;
      
        // Track the longest palindrome prefix
        int longestPalindromePrefixLength = 0;
        int stringLength = s.length();
      
        // Iterate through each character to find longest palindrome prefix
        for (int i = 0; i < stringLength; ++i) {
            // Convert character to numeric value (1-26 for 'a'-'z')
            int charValue = s.charAt(i) - 'a' + 1;
          
            // Update prefix hash: hash = hash * base + charValue
            prefixHash = (int) (((long) prefixHash * BASE + charValue) % MOD);
          
            // Update suffix hash: hash = hash + charValue * base^i
            suffixHash = (int) ((suffixHash + (long) charValue * multiplier) % MOD);
          
            // Update multiplier for next iteration
            multiplier = (int) (((long) multiplier * BASE) % MOD);
          
            // If hashes match, we found a palindrome prefix
            if (prefixHash == suffixHash) {
                longestPalindromePrefixLength = i + 1;
            }
        }
      
        // If entire string is already a palindrome, return as is
        if (longestPalindromePrefixLength == stringLength) {
            return s;
        }
      
        // Build result by reversing the non-palindrome suffix and prepending it
        String nonPalindromeSuffix = s.substring(longestPalindromePrefixLength);
        String reversedSuffix = new StringBuilder(nonPalindromeSuffix).reverse().toString();
      
        return reversedSuffix + s;
    }
}

File: 0214-shortest-palindrome/0214-shortest-palindrome.ts
/**
 * @param s 문자열
 * @returns 가장 짧은 팰린드롬
 */
function shortestPalindrome(s: string): string {
    // 문자열 길이가 0 또는 1이면 이미 팰린드롬
    if (s.length <= 1) {
        return s;
    }

    // 문자열 뒤집기
    const reversed: string = s
        .split('')
        .reverse()
        .join('');

    // 원본 + 구분자 + 뒤집은 문자열
    const combined: string = s + '#' + reversed;

    // KMP의 LPS(Longest Prefix Suffix) 배열
    const lps: number[] = new Array(combined.length).fill(0);

    // LPS 배열 생성
    for (let i: number = 1; i < combined.length; i++) {
        let length: number = lps[i - 1];

        // 현재 문자가 일치하지 않으면 이전 LPS 위치로 이동
        while (
            length > 0 &&
            combined[i] !== combined[length]
        ) {
            length = lps[length - 1];
        }

        // 문자가 일치하면 길이 증가
        if (combined[i] === combined[length]) {
            length++;
        }

        lps[i] = length;
    }

    // 가장 긴 팰린드롬 Prefix의 길이
    const palindromeLength: number =
        lps[combined.length - 1];

    // 팰린드롬에 포함되지 않은 뒷부분
    const suffix: string = s.substring(palindromeLength);

    // 뒷부분을 뒤집어서 앞에 추가
    const prefixToAdd: string = suffix
        .split('')
        .reverse()
        .join('');

    return prefixToAdd + s;
}